text/0000-let-else.md

```diff
@@ -0,0 +1,602 @@
+which is considered to be idiomatic rust.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+None known at time of writing due to extensive pre-discussion in Zulip:
+https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/.60let.20pattern.20.3D.20expr.20else.20.7B.20.2E.2E.2E.20.7D.60.20statements
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## if-let-chains
+
+An RFC exists for a (unimplemented at time of writing) feature called [if-let-chains][]:
+
+```rust
+if let Some(foo) = expr() && foo.is_baz() && let Ok(yay) = qux(foo) { ... }
+```
+
+While this RFC does not introduce or propose the same thing for let-else it attempts to allow it to be a future possibility for
+potential future consistency with if-let-chains.
+
+The primary obstacle is existing operator order precedence.
+Given the above example, it would likely be parsed as follows with ordinary operator precedence rules for `&&`:
+```rust
+let Some(foo) = (expr() && foo.is_baz() && let Ok(yay) = qux(foo) else { ... })
+```
+
+However, given that all existing occurrences of this behavior before this RFC are type errors anyways,
+a specific boolean-only case can be avoided and thus parsing can be changed to leave the door open to this possible extension.
+This boolean case is always equivalent to a less flexible `if` statement and as such is not useful.
+
+```rust
+let maybe = Some(2);
+let has_thing = true;
+
+// Always an error regardless, because && only operates on booleans.
+let Some(x) = maybe && has_thing else {
+    return;
+};
+```
+
+```rust
+let a = false;
+let b = false;
+
+// The RFC proposes boolean patterns with a lazy boolean operator (&& or ||)
+//  be made into a compile error, for potential future compatibility with if-let-chains.
+let true = a && b else {
+    return;
+};
+```
+
+Note also that this does not work today either, because booleans are refutable patterns:
+```
+error[E0005]: refutable pattern in local binding: `false` not covered
+ --> src/main.rs:5:9
+  |
+5 |     let true = a && b;
+  |         ^^^^ pattern `false` not covered
+  |
+  = note: `let` bindings require an "irrefutable pattern", like a `struct` or an `enum` with only one variant
+```
+
+## Fall-back assignment
+
+This RFC does not suggest that we do any of these, but notes that they would be future possibilities.
+
+If fall-back assignment as discussed above in [rationale-and-alternatives][] is desirable, it could be added with an additional keyword as a future extension:
+
+```rust
+enum AnEnum {
+    Variant1(u32),
+    Variant2(String),
+}
+
+let AnEnum::Variant1(a) = x else assign a {
+    a = 42;
+};
+```
+
+Another potential form of the fall-back extension:
+
+```rust
+let Ok(a) = x else match {
+    Err(e) => return Err(e.into()),
+}
+```
+
+## let-else within if-let
+
+Conceivable. This RFC makes no judgement, even if the author does.
```

Comment:
I would prefer to see this rejected; it seems entirely too confusing. We could always decide to allow it later if we *really* want to, but let's not just default to "sure, why not".

Reply:
`let x = y` is a statement, not an expression. So the same should be true for `let`-`else`. Even with if-let-chains, a block would be expected where `else` is.